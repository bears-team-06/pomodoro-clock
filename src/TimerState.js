const TimerState = {
    Stopped: 0,
    Running: 1,
    Paused: 2
};

export default TimerState