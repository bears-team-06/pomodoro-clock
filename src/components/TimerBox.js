import React, {Component} from "react";
import TimerState from "../TimerState"

class TimerBox extends Component {
    componentWillReceiveProps(newProps) {
        if(this.state.timerState !== TimerState.Running && newProps.timerState === TimerState.Running) {
            this.startTicking()
        }
        const hasStateChangedToPause = this.state.timerState === TimerState.Running && newProps.timerState === TimerState.Paused
        const hasStateChangedToStopped = this.state.timerState !== TimerState.Stopped && newProps.timerState === TimerState.Stopped
        const hasSessionTimeChanged = this.props.sessionTime !== newProps.sessionTime 
        if(hasStateChangedToPause || hasStateChangedToStopped) {
            this.stopTimer()
        }
        if(hasSessionTimeChanged || hasStateChangedToStopped) {
            this.setState({
                timeInSeconds: newProps.sessionTime
            })
        }
        this.setState({
            timerState: newProps.timerState
        })
    }

    constructor(props) {
        super(props);
        this.state = {
            timeInSeconds: this.props.sessionTime,
            timerState: this.props.timerState
        };
        this.timer = null
    }

    get minutes() {
        let minutes = parseInt(this.state.timeInSeconds / 60);
        return minutes > 9 ? `${minutes}` : `0${minutes}`;
    }

    get seconds() {
        let seconds = parseInt(this.state.timeInSeconds % 60);
        return seconds > 9 ? `${seconds}` : `0${seconds}`;
    }

    stopTimer() {
        this.timer && clearInterval(this.timer)
    }

    onTimerComplete() {
        this.stopTimer()
        this.props.onTimerComplete && this.props.onTimerComplete()
    }

    startTicking = () => {
        this.timer = setInterval(() => {
            if (this.state.timeInSeconds === 0) {
                this.onTimerComplete()
            } else {
                this.setState((state, props) => ({
                    timeInSeconds: state.timeInSeconds - 1
                }));
            }
        }, 1000);
    }

    render() {
        return (
            <div>
                <span>{this.props.sessionName}</span>
                <span>{`${this.minutes}:${this.seconds}`}</span>
            </div>
        );
    }
}

export default TimerBox;
