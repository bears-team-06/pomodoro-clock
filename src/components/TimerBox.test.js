import React from 'react'
import { shallow } from 'enzyme'
import TimerState from "../TimerState";
import TimerBox from "./TimerBox";

describe('Timer state', () => {

    it('renders snapshot', () => {
        let timerBox = shallow(<TimerBox sessionTime={1600} timerState={TimerState.Running} sessionName={"Short Break"}/>)
        expect(timerBox).toMatchSnapshot();
    })

    it('calls starts set Interval for 1 second when starts ticking', () => {
        jest.useFakeTimers();
        const timerBox = shallow(<TimerBox sessionTime={1600} timerState={TimerState.Running} sessionName={"Short Break"} />)
        timerBox.instance().startTicking()
        expect(setInterval).toHaveBeenCalledTimes(1);
        expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);
    })

    it('does not start timer when timerstate is not changed', () => {
        const timerBox = shallow(<TimerBox sessionTime={10} timerState={TimerState.Stopped} sessionName={"Short Break"}/>)
        const startTickingSpy = jest.spyOn(timerBox.instance(), 'startTicking')
        timerBox.setProps({
            timerState: TimerState.Stopped
        })
        expect(startTickingSpy).toHaveBeenCalledTimes(0)
    })

    it('start timer when timerstate is changed to running from stopped', () => {
        const timerBox = shallow(<TimerBox sessionTime={10} timerState={TimerState.Stopped} sessionName={"Short Break"}/>)
        const startTickingSpy = jest.spyOn(timerBox.instance(), 'startTicking')
        timerBox.setProps({
            timerState: TimerState.Running
        })
        expect(startTickingSpy).toHaveBeenCalledTimes(1)
    })

    it('stops timer when state is changed to pause', () => {
        const timerBox = shallow(<TimerBox sessionTime={10} timerState={TimerState.Running} sessionName={"Short Break"}/>)
        const stopTimerSpy = jest.spyOn(timerBox.instance(), 'stopTimer')
        timerBox.setProps({
            timerState: TimerState.Paused
        })
        expect(stopTimerSpy).toHaveBeenCalledTimes(1)
    })

    it('stops timer when state is changed to stopped', () => {
        const timerBox = shallow(<TimerBox sessionTime={10} timerState={TimerState.Running} sessionName={"Short Break"}/>)
        const stopTimerSpy = jest.spyOn(timerBox.instance(), 'stopTimer')
        timerBox.setProps({
            timerState: TimerState.Stopped
        })
        expect(stopTimerSpy).toHaveBeenCalledTimes(1)
    })

    it('calls on timerComplete when session time changes to zero', () => {
        jest.useFakeTimers();
        const onTimerComplete = jest.fn()
        const timerBox = shallow(<TimerBox sessionTime={1} timerState={TimerState.Running} sessionName={"Short Break"} onTimerComplete={onTimerComplete}/>)
        expect(onTimerComplete).toHaveBeenCalledTimes(0)
        const playAlarmSpy = jest.spyOn(timerBox.instance(), 'playAlarm');
        timerBox.instance().startTicking()
        jest.runAllTimers()
        expect(setInterval).toHaveBeenCalledTimes(1);
        expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);
        expect(onTimerComplete).toHaveBeenCalledTimes(1)
        expect(playAlarmSpy).toHaveBeenCalledTimes(1)
    })

})