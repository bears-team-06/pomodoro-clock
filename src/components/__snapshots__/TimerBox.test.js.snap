// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer state renders snapshot 1`] = `
<div>
  <span
    id="timer-box-name"
  >
    Short Break
  </span>
  <span
    id="timer-box-time"
  >
    26 : 40
  </span>
</div>
`;
